src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinActivityConversionExtensions.cs

```diff
@@ -0,0 +1,219 @@
+
+            var activitySource = activity.Source;
+            if (!string.IsNullOrEmpty(activitySource.Name))
+            {
+                PooledList<KeyValuePair<string, string>>.Add(ref attributeEnumerationState.Tags, new KeyValuePair<string, string>("library.name", activitySource.Name));
+                if (!string.IsNullOrEmpty(activitySource.Version))
+                {
+                    PooledList<KeyValuePair<string, string>>.Add(ref attributeEnumerationState.Tags, new KeyValuePair<string, string>("library.version", activitySource.Version));
+                }
+            }
+
+            var localEndpoint = defaultLocalEndpoint;
+
+            var serviceName = attributeEnumerationState.ServiceName;
+
+            // override default service name
+            if (!string.IsNullOrWhiteSpace(serviceName))
+            {
+                if (!string.IsNullOrWhiteSpace(attributeEnumerationState.ServiceNamespace))
+                {
+                    serviceName = attributeEnumerationState.ServiceNamespace + "." + serviceName;
+                }
+
+                if (!LocalEndpointCache.TryGetValue(serviceName, out localEndpoint))
+                {
+                    localEndpoint = defaultLocalEndpoint.Clone(serviceName);
+                    LocalEndpointCache.TryAdd(serviceName, localEndpoint);
+                }
+            }
+
+            ZipkinEndpoint remoteEndpoint = null;
+            if ((activity.Kind == ActivityKind.Client || activity.Kind == ActivityKind.Producer) && attributeEnumerationState.RemoteEndpointServiceName != null)
+            {
+                remoteEndpoint = RemoteEndpointCache.GetOrAdd(attributeEnumerationState.RemoteEndpointServiceName, ZipkinEndpoint.Create);
+            }
+
+            var annotations = PooledList<ZipkinAnnotation>.Create();
+            ListEnumerator<ActivityEvent, PooledList<ZipkinAnnotation>>.AllocationFreeForEach(activity.Events, ref annotations, ProcessActivityEventsRef);
+
+            return new ZipkinSpan(
+                EncodeTraceId(context.TraceId, useShortTraceIds),
+                parentId,
+                EncodeSpanId(context.SpanId),
+                ToActivityKind(activity),
+                activity.OperationName,
+                activity.StartTimeUtc.ToEpochMicroseconds(),
+                duration: (long)activity.Duration.TotalMilliseconds * 1000, // duration in Microseconds
+                localEndpoint,
+                remoteEndpoint,
+                annotations,
+                attributeEnumerationState.Tags,
+                null,
+                null);
+        }
+
+        internal static string EncodeSpanId(ActivitySpanId spanId)
+        {
+            return spanId.ToHexString();
+        }
+
+        internal static long ToEpochMicroseconds(this DateTimeOffset timestamp)
+        {
+            return timestamp.ToUnixTimeMilliseconds() * 1000;
```

Comment:
Same here, it seems we're losing precision.

Reply:
Fixed